pkg/info/info.go

```diff
@@ -26,53 +26,99 @@ type InfoServer struct {
 
 func NewInfoServer(
 	component string,
-	options ...func(*InfoServer),
+	options ...ServerOption,
 ) *InfoServer {
 	srv := &InfoServer{
 		component: component,
 	}
 
 	for _, o := range options {
-		o(srv)
+		o.applyServerOption(srv)
 	}
 
 	return srv
 }
 
-func WithLabelSet(getLabelSet func() []labelpb.ZLabelSet) func(*InfoServer) {
-	return func(s *InfoServer) {
-		s.getLabelSet = getLabelSet
+type ServerOption interface {
+	applyServerOption(*InfoServer)
+}
+
+type serverOptionFunc func(*InfoServer)
+
+func (fn serverOptionFunc) applyServerOption(cfg *InfoServer) {
+	fn(cfg)
+}
+
+func WithLabelSetFunc(getLabelSet ...func() []labelpb.ZLabelSet) ServerOption {
+	if len(getLabelSet) == 0 {
+		return serverOptionFunc(func(s *InfoServer) {
+			s.getLabelSet = func() []labelpb.ZLabelSet { return []labelpb.ZLabelSet{} }
+		})
 	}
+
+	return serverOptionFunc(func(s *InfoServer) {
+		s.getLabelSet = getLabelSet[0]
+	})
 }
 
-func WithStoreInfo(getStoreInfo func() *infopb.StoreInfo) func(*InfoServer) {
-	return func(s *InfoServer) {
-		s.getStoreInfo = getStoreInfo
+func WithStoreInfoFunc(getStoreInfo ...func() *infopb.StoreInfo) ServerOption {
+	if len(getStoreInfo) == 0 {
+		return serverOptionFunc(func(s *InfoServer) {
+			s.getStoreInfo = func() *infopb.StoreInfo { return &infopb.StoreInfo{} }
+		})
 	}
+
+	return serverOptionFunc(func(s *InfoServer) {
+		s.getStoreInfo = getStoreInfo[0]
+	})
```

Comment:
I like this pattern :100: 

The only thing that might be a con is the function signature allows more than 1 arguments, and it is not clear to the users that anything other than the first argument will be ignored. Maybe we can add a comment on this function saying that.

Reply:
That's a good suggestion, I wanted to add more comments overall in the info package, I'll do that.